### ISSUE TITLE
Rust 1.18 regression: coinbaser 0.1.0

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#40955]
Reported on [cargobomb](https://brson.github.io/cargobomb-reports/stable-v-nightly-2017-03-29/res/nightly/reg/coinbaser-0.1.0/log.txt)

```
boom! running `docker start -a fcd992283fd7ae3f00c77dc9c371733b58ef79cf654f4113f69b8555e73a9612`
kablam!    Compiling coinbaser v0.1.0 (file:///source)
kablam! error[E0432]: unresolved import `hyper::client::Client`
kablam!  --> src/coinbaser.rs:5:5
kablam!   |
kablam! 5 | use hyper::client::Client;
kablam!   |     ^^^^^^^^^^^^^^^^^^^^^ Maybe a missing `extern crate hyper;`?
kablam! 
kablam! error[E0432]: unresolved import `hyper::status::StatusCode`
kablam!  --> src/coinbaser.rs:6:5
kablam!   |
kablam! 6 | use hyper::status::StatusCode;
kablam!   |     ^^^^^^^^^^^^^^^^^^^^^^^^^ Maybe a missing `extern crate hyper;`?
kablam! 
kablam! error[E0432]: unresolved import `hyper::Url`
kablam!  --> src/coinbaser.rs:7:5
kablam!   |
kablam! 7 | use hyper::Url;
kablam!   |     ^^^^^^^^^^ Maybe a missing `extern crate hyper;`?
kablam! 
kablam! error[E0432]: unresolved import `error::Error`
kablam!  --> src/coinbaser.rs:9:5
kablam!   |
kablam! 9 | use error::Error as _CoinError;
kablam!   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ Did you mean `std::error`?
kablam! 
kablam! error: main function not found
kablam! 
kablam! error: aborting due to 5 previous errors
kablam! 
kablam! error: Could not compile `coinbaser`.
kablam! 
```

Unfortunately the source is not available, but @patrickjm can you perhaps help reproduce?

**Comment by jethrogb** [comment]
What do you mean source is not available? How did you build it in the first place?

**Comment by brson** [comment]
@jethrogb I mean that I can't find the git repo. The source can be acquired from the published crate tarball.

**Comment by patrickjm** [comment]
The repo has been deleted for some time now - I tried to delete it from cargo but that's not allowed.

**Comment by brson** [comment]
Should probably still be investigated, unless @patrickjm doesn't want us testing his yanked crate. We'll have to grab it from crates.io.

I wonder if maybe cargobomb shouldn't test yanked crates.

**Comment by alexcrichton** [comment]
This is a regression in Cargo. With a `Cargo.toml` that looks like:

```
[project]
name = "foo"

[[bin]]
name = "foo"
```

And files at `src/main.rs` and `src/foo.rs` on stable Cargo will compile `src/main.rs`, on beta Cargo will compile `src/foo.rs` instead. 

cc @jmatraszek, is this related to https://github.com/rust-lang/cargo/pull/3609 perhaps?

**Comment by jmatraszek** [comment]
Hi @alexcrichton, will investigate this today, but looks like it is related.

**Comment by alexcrichton** [comment]
Thanks @jmatraszek!

**Comment by patrickjm** [comment]
Sorry about the hassle guys. I removed it because I didn't want my useless and dead project polluting things up, but that just caused more trouble!

No worries about testing my yanked crate. @brson

**Comment by alexcrichton** [comment]
@patrickjm oh no worries! This crate helped us detect a regression and what's what we're worried about :)

**Comment by jmatraszek** [comment]
Hi @alexcrichton,
I think this was caused by this: https://github.com/rust-lang/cargo/blob/master/src/cargo/util/toml.rs#L1367, but it worked because of the bug I fixed in 3609 (cargo always assumed `src/main.rs` as a source file for a bin target).

You can find the proposed fix in rust-lang/cargo#3898. I tried to get rid of quick fixes and implement a full featured solution for inferring bin target's source file paths. Please review the inferring logic — I am not 100% sure that it covers all cargo's usecases, but it seems reasonable for me.

**Comment by alexcrichton** [comment]
Awesome thanks for the investigation @jmatraszek!
